runtime/src/installed_scheduler_pool.rs

```diff
@@ -0,0 +1,544 @@
+///          InstalledSchedulerPool{{InstalledSchedulerPool}};
+///      end
+///
+///      subgraph solana-ledger
+///          ExecuteBatch(["execute_batch()"]);
+///      end
+///
+///      subgraph solana-scheduler-pool
+///          SchedulerPool;
+///          PooledScheduler;
+///      end
+///
+///      subgraph solana-scheduler
+///          WithSchedulingMode{{WithSchedulingMode}};
+///      end
+///
+///      SchedulingContext -- refs --> Bank;
+///      BankForks -- owns --> Bank;
+///      BankForks -- owns --> InstalledSchedulerPool;
+///      Bank -- refs --> InstalledScheduler;
+///
+///      SchedulerPool -. impls .-> InstalledSchedulerPool;
+///      PooledScheduler -. impls .-> InstalledScheduler;
+///      InstalledScheduler -- refs --> SchedulingContext;
+///      SchedulingContext -. impls .-> WithSchedulingMode;
+///
+///      PooledScheduler -- refs --> SchedulerPool;
+///      SchedulerPool -- owns --> PooledScheduler;
+///      PooledScheduler -. calls .-> ExecuteBatch;
+///
+///      solana-scheduler-pool -- deps --> solana-scheduler;
+///      solana-scheduler-pool -- deps --> solana-ledger;
+///      solana-scheduler-pool -- deps --> solana-runtime;
+///      solana-ledger -- deps --> solana-runtime;
+///      solana-runtime -- deps --> solana-scheduler;
+///  ```
+#[cfg_attr(any(test, feature = "test-in-workspace"), automock)]
+// suppress false clippy complaints arising from mockall-derive:
+//   warning: `#[must_use]` has no effect when applied to a struct field
+//   warning: the following explicit lifetimes could be elided: 'a
+#[allow(unused_attributes, clippy::needless_lifetimes)]
+// Send + Sync is needed to be a field of Bank
+pub trait InstalledScheduler<SEA: ScheduleExecutionArg>: Send + Sync + Debug {
+    fn id(&self) -> SchedulerId;
+    fn pool(&self) -> InstalledSchedulerPoolArc;
+
+    // Calling this is illegal as soon as schedule_termiantion is called on &self.
+    fn schedule_execution<'a>(&'a self, transaction_with_index: SEA::TransactionWithIndex<'a>);
+
+    #[must_use]
+    fn wait_for_termination(&mut self, reason: &WaitReason) -> Option<ResultWithTimings>;
+
+    fn context<'a>(&'a self) -> Option<&'a SchedulingContext>;
+    fn replace_context(&mut self, context: SchedulingContext);
+}
+
+pub type InstalledSchedulerPoolArc = Arc<dyn InstalledSchedulerPool>;
+
+pub type SchedulerId = u64;
+
+pub type ResultWithTimings = (Result<()>, ExecuteTimings);
+
+pub trait WithTransactionAndIndex: Send + Sync + Debug {
+    fn with_transaction_and_index(&self, callback: impl FnOnce(&SanitizedTransaction, usize));
+}
+
+impl<
+        T: Send + Sync + Debug + Borrow<SanitizedTransaction>,
+        U: Send + Sync + Debug + Borrow<usize>,
+        Z: Send + Sync + Debug + Deref<Target = (T, U)>,
+    > WithTransactionAndIndex for Z
+{
+    fn with_transaction_and_index(&self, callback: impl FnOnce(&SanitizedTransaction, usize)) {
+        callback(self.0.borrow(), *self.1.borrow());
+    }
+}
```

Comment:
>  And I need to read the implementation, which is non-trivial, considering all the type dance, to find out that it is just to destructure a tuple :|

hehe, i feel the same way. although i omitted explicit documentation, but I'm envisioning to introduce full-fledged struct named like `Arc<ExecutionEnvironment>` or `Arc<SchedulableTask>` (better name is welcome!), which contains `SanitizedTransaction` along with some scheduling/execution states for the transaction. also, it will also abstract over normal Transaction/Bundle.

Reply:
I do not have a full model of this branch code in my head.
And I must have forgotten a good chunk of what I knew before my vacation %)

But, are you sure it is necessary still?  Even if that `ExecutionEnvironment` is added.
I mean, if it contains a transaction and an index, the caller of the `handle()` method could just extract it and pass a tuple into the `handle()` call.
This caller could be a new implementation of the `ScheduledTransactionHandler` that extracts these values and then calls `DefaultTransactionHandler`.

As long as `DefaultTransactionHandler::handle()` does not need any additional functionality, it seems best to reflect that in the `handle()` (and `DefaultTransactionHandler`).
I think it is called a principle of the least dependency or something like that.
If you do not need something, do not require it - the best default.
It provides for the most flexibility for the current and any future callers.
I might be missing something, but if the above is correct, future extensions of the callers do not matter.  Only future plans for the `DefaultTransactionHandler::handle()` implementation matter.

And, unless you are really planning this functionality extension to happen immediately (which could absolutely be the case), still consider avoiding writing code for that.  As things may change between now and when that code is added.   And it could be added in a different manner.
It is always possible to refactor.  And changing simple to more complex is easier than the other way around :)

Reply:
> Only future plans for the `DefaultTransactionHandler::handle()` implementation matter.
> ...
> I think it is called a principle of the least dependency or something like that.
If you do not need something, do not require it - the best default.
It provides for the most flexibility for the current and any future callers.
> ...
> And, unless you are really planning this functionality extension to happen immediately (which could absolutely be the case), still consider avoiding writing code for that. As things may change between now and when that code is added. And it could be added in a different manner.
It is always possible to refactor. And changing simple to more complex is easier than the other way around :)

yeah, I'm planning to extend `ScheduledTransactionHandler::handle()` to make use of  upcoming `ScheduledTask`/`ExecutionEnvironment` _soon_ (hopefully!). that said, you have a good point (as usual with all other review comments!) so I reverted `::handle()` signature a bit for now: df692d77e750fc1b5b73f63622655ba3363ab4ea

> With the `ScheduledTransactionHandler` becoming only parametric over lifetimes,
> removing the `SEA` type argument:

however, note that removing SEA from the handler trait isn't possible anymore due to 1c89b9a0b5263a0cc395eaaa76e79662806131ce. i needed to parameterize `SchedulerPool` and friends including the handler trait to add tests. I wanted to defer this intense type dance session in similar reasoning; but seems not anymore.. ;p